implementations/rust/ockam/ockam_api/src/session/mod.rs

```diff
@@ -228,3 +228,97 @@ impl Worker for Collector {
+
+    let medic_task: tokio::task::JoinHandle<std::result::Result<(), Error>> =
+        tokio::spawn(medic.start(new_ctx));
+
+    // Medic relies on echo to verify if a session is alive
+    ctx.start_worker(Address::from_string("echo"), Echoer).await?;
+
+    // Hop serves as simple neutral address we can use
+    ctx.start_worker(Address::from_string("hop"), Hop).await?;
+
+    let replacer_called = Arc::new(AtomicBool::new(false));
+    let replacer_can_return = Arc::new(AtomicBool::new(false));
+
+    {
+        let mut session = Session::new(route!["broken_route"]);
+        let replacer_called = replacer_called.clone();
+        let replacer_can_return = replacer_can_return.clone();
+        session.set_replacer(Box::new(move |_| {
+            let replacer_called = replacer_called.clone();
+            let replacer_can_return = replacer_can_return.clone();
+            Box::pin(async move {
+                log::info!("replacer called");
+                replacer_called.store(true, Ordering::Release);
+                while !replacer_can_return.load(Ordering::Acquire) {
+                    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+                }
+                // en empty route would do the trick, but a hop is more realistic
+                Ok(route!["hop"])
+            })
+        }));
+
+        sessions.lock().unwrap().add(session);
+    }
+
+    {
+        // Initially it's up
+        let mut guard = sessions.lock().unwrap();
+        let (_, session) = guard.iter_mut().next().unwrap();
+        assert_eq!(session.status(), Status::Up);
+        assert_eq!(session.ping_route(), &route!["broken_route"]);
+    }
+
+    // Since the route is broken eventually it will be degraded and will call the replacer
+    while !replacer_called.load(Ordering::Acquire) {
+        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+    }
+
+    {
+        // Check the session is now marked as degraded
+        let guard = sessions.lock().unwrap();
+        let (_, session) = guard.iter().next().unwrap();
+        assert_eq!(session.status(), Status::Degraded);
+        assert_eq!(session.ping_route(), &route!["broken_route"]);
+    }
+
+    // Now we allow the replacer to return and replace the route
+    replacer_can_return.store(true, Ordering::Release);
+
+    loop {
+        {
+            // Check that the session is now up, since we don't have any
+            // synchronization we keep to keep checking until it's up
+            let guard = sessions.lock().unwrap();
+            let (_, session) = guard.iter().next().unwrap();
+            if session.status() == Status::Up {
+                assert_eq!(session.ping_route(), &route!["hop"]);
+                break;
+            }
+        }
+
+        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+        continue;
+    }
+
+    // Shut down the test
+    medic_task.abort();
+    ctx.stop().await?;
+    Ok(())
+}
```

Comment:
Make sure to run `cargo fmt`. I think there are a couple of formatting issues that should be fixed.